js/h5p-tooltip.js

```diff
@@ -0,0 +1,149 @@
+
+    triggeringElement.appendChild(tooltip);
+
+    // Aria-describedby will override aria-hidden
+    if (!options.ariaHidden) {
+      triggeringElement.setAttribute('aria-describedby', tooltipId);
+    }
+    
+    //Add event listeners to triggeringElement
+    triggeringElement.onmouseenter = function () {
+      tooltip.classList.add('h5p-tooltip-visible');
+      self.ensureTooltipVisible();
+      self.hover = true;
+      self.listenForEscape();
+    }
+    triggeringElement.onmouseleave = function () {
+      // Only hide tooltip if not focused
+      if(self.focus === false) {
+        tooltip.classList.remove('h5p-tooltip-visible');
+        self.stopListeningForEscape();
+      }
+      self.hover = false;
+    }
+    triggeringElement.addEventListener('focusin', function () {
+      tooltip.classList.add('h5p-tooltip-visible');
+      self.ensureTooltipVisible();
+      self.focus = true;
+      self.listenForEscape();
+    });
+    triggeringElement.addEventListener('focusout', function () {
+      // Only hide tooltip if not hovered
+      if(self.hover === false) {
+        tooltip.classList.remove('h5p-tooltip-visible');
+        self.stopListeningForEscape();
+      }
+      self.focus = false;
+    });
+
+    /**
+     * Ensures that all of the tooltip is visible
+     */
+    this.ensureTooltipVisible = function () {
+      const availableWidth = document.body.clientWidth;
+      const tooltipWidth = tooltip.offsetWidth;
+      const triggeringElementWidth = triggeringElement.clientWidth;
+      const triggeringElementOffsetLeft = triggeringElement.offsetLeft;
+
+      // Going out of screen on left side
+      if (triggeringElementOffsetLeft + triggeringElementWidth < tooltipWidth) {
+        tooltip.classList.add('h5p-tooltip-left');
+        tooltip.classList.remove('h5p-tooltip-right');
+      }
+      // Going out of screen on right side
+      else if ((triggeringElementOffsetLeft + tooltipWidth) > availableWidth) {
+        tooltip.classList.add('h5p-tooltip-right');
+        tooltip.classList.remove('h5p-tooltip-left');
+      }
+      else {
+        tooltip.classList.remove('h5p-tooltip-right');
+        tooltip.classList.remove('h5p-tooltip-left');
+      }
+    };
+
+    /**
+     * Add listener to iframe body, as esc keypress would not be detected otherwise 
+     */
+    this.listenForEscape = function () {
+      iframeBody.addEventListener('keydown', escapeFunction, true);
+    }
+
+    /**
+     * Remove iframe body listener
+     */
+     this.stopListeningForEscape = function () {
+      iframeBody.removeEventListener('keydown', escapeFunction, true);
+    }
+    
+    /**
+     * Retrieve tooltip
+     *
+     * @return {HTMLElement}
+     */
+    this.getElement = function () {
+      return tooltip;
+    };
+  }
+
+  Tooltip.prototype = Object.create(EventDispatcher.prototype);
+  Tooltip.prototype.constructor = Tooltip;
+
+  return Tooltip;
+
+}(H5P.EventDispatcher));
+
+H5P.Tooltip.uniqueId = -1;
```

Comment:
In general, your code is very tidy and easy to read :star: 
I have one general improval suggestion, which is about avoiding duplication of code. I.e: I would at least create a function for closing and opening the tooltip, which should handle the adding/removing the h5p-tooltip-visible class and setting up/removing event listening.

Reply:
Another thing:
When you put functions on this (this.listenForEscape) you make them "public". Is that intentional? If not, you should make them private, by doing something like:
`const listenForEscape = function () {`
If you create a function for opening the tooltip, it might be this function is only invoked once. Then you don't need a separate function for it at all...

Reply:
It was not intentional to make it public, thanks for the heads up ^^